File: backend/app/models/stitching.py
"""
Pydantic модели для склейки планов (stitching)
"""

from typing import List, Optional, Tuple, Literal
from pydantic import BaseModel, Field, field_validator


class TransformInput(BaseModel):
    """Affine transformation parameters."""
    translate_x: float
    translate_y: float
    scale_x: float = Field(gt=0, description="Must be > 0")
    scale_y: float = Field(gt=0, description="Must be > 0")
    rotation_deg: float = Field(ge=0, le=360)


class ClipPolygonInput(BaseModel):
    """Polygon for clipping (subtract operation)."""
    type: Literal["subtract"]
    points: List[Tuple[float, float]] = Field(min_length=3)


class RectCropInput(BaseModel):
    """Rectangular crop in image space."""
    x: float = Field(ge=0)
    y: float = Field(ge=0)
    width: float = Field(gt=0)
    height: float = Field(gt=0)


class SourcePlanInput(BaseModel):
    """Single source plan with transformations."""
    reconstruction_id: str
    transform: TransformInput
    clip_polygons: List[ClipPolygonInput]
    rect_crop: Optional[RectCropInput] = None
    image_width_px: int = Field(gt=0)
    image_height_px: int = Field(gt=0)
    z_index: int = Field(ge=0)


class StitchingRequest(BaseModel):
    """Request to stitch multiple plans."""
    name: str = Field(min_length=1, max_length=255)
    building_id: str  # UUID validation can be added via validator
    floor_number: int = Field(ge=0)
    source_plans: List[SourcePlanInput] = Field(min_length=2)

    @field_validator('source_plans')
    @classmethod
    def validate_source_plans_count(cls, v: List[SourcePlanInput]) -> List[SourcePlanInput]:
        """Ensure at least 2 source plans are provided."""
        if len(v) < 2:
            raise ValueError('At least 2 source plans are required for stitching')
        return v


class StitchingResponse(BaseModel):
    """Response after stitching."""
    id: int
    name: str
    status: int
    source_reconstruction_ids: List[int]
    building_id: str
    floor_number: int
    rooms_count: int
    walls_count: int
    warnings: Optional[List[str]] = None

File: frontend/src/types/stitching.ts
// Transform
export interface Transform {
  translate_x: number;
  translate_y: number;
  scale_x: number;
  scale_y: number;
  rotation_deg: number;
}

// Clip polygon
export interface ClipPolygon {
  type: "subtract";
  points: [number, number][];
}

// Rect crop
export interface RectCrop {
  x: number;
  y: number;
  width: number;
  height: number;
}

// Source plan input
export interface SourcePlanInput {
  reconstruction_id: string;
  transform: Transform;
  clip_polygons: ClipPolygon[];
  rect_crop: RectCrop | null;
  image_width_px: number;
  image_height_px: number;
  z_index: number;
}

// Stitching request
export interface StitchingRequest {
  name: string;
  building_id: string;
  floor_number: number;
  source_plans: SourcePlanInput[];
}

// Stitching response
export interface StitchingResponse {
  id: number;
  name: string;
  status: number;
  source_reconstruction_ids: number[];
  building_id: string;
  floor_number: number;
  rooms_count: number;
  walls_count: number;
  warnings?: string[];
}

// Reconstruction list item (for selection)
export interface ReconstructionListItem {
  id: number;
  name: string;
  preview_url: string | null;
  rooms_count: number;
  walls_count: number;
  created_at: string;
  rotation_angle: number;
  building_id: string | null;
  floor_number: number | null;
  status?: string;
}

// Layer data (canvas state)
export interface LayerData {
  reconstructionId: string;
  name: string;
  originalImageUrl: string;
  previewUrl: string;
  vectorModel: VectorModel;
  transform: Transform;
  clipPolygons: ClipPolygon[];
  rectCrop: RectCrop | null;
  imageWidth: number;
  imageHeight: number;
  zIndex: number;
  color: string;
  maskOpacity: number;
  showMask: boolean;
}

// Vector model (from reconstruction)
export interface VectorModel {
  walls: Wall[];
  rooms: Room[];
  doors: Door[];
}

export interface Wall {
  id: string;
  points: Point2D[];
  thickness: number;
}

export interface Room {
  id: string;
  name: string;
  polygon: Point2D[];
  center: Point2D;
  room_type: string;
}

export interface Door {
  id: string;
  position: Point2D;
  width: number;
}

export interface Point2D {
  x: number;
  y: number;
}

// Stitching state
export interface StitchingState {
  step: 1 | 2;
  selectedReconstructionIds: string[];
  buildingId: string;
  floorNumber: number;
  layers: LayerData[];
  activeTool: "move" | "rotate" | "rect_crop" | "polygon_clip";
  selectedLayerId: string | null;
  isLoading: boolean;
  error: string | null;
}

// Snapshot for undo/redo
export interface StitchingSnapshot {
  layers: LayerSnapshot[];
}

export interface LayerSnapshot {
  reconstructionId: string;
  transform: Transform;
  clipPolygons: ClipPolygon[];
  rectCrop: RectCrop | null;
  zIndex: number;
}
